### ISSUE TITLE
Questions about _train_epoch in trainer.py

### ISSUE THREAD

**Issue by ciuzaak** [Jia-Research-Lab/Context-Aware-Consistency#11]
`        if self.mode == 'supervised':`
`            # dataloader = iter(self.supervised_loader)`
`            # tbar = tqdm(range(len(self.supervised_loader)), ncols=135)`
`            dataloader = iter(cycle(self.supervised_loader))`
`            tbar = tqdm(range(self.iter_per_epoch), ncols=135)`
`        else:`
`            dataloader = iter(zip(cycle(self.supervised_loader), cycle(self.unsupervised_loader)))`
`            tbar = tqdm(range(self.iter_per_epoch), ncols=135)`

The comment part is your original code. 
In the semi-supervised method, 'cycle' was used to expand the number of iterations of labeled images. Obviously the number of iterations in the fully-supervised is much less. I think this comparison may be unfair. What is your opinion or modification plan? Looking forward  to your answer, thanks!

**Comment by X-Lai** [comment]
Thank you for raising this question.

Yes, you are correct at first. Looks like the epoch size is `len(self.supervised_loader)` rather than `self.iter_per_epoch`, i.e., 1157 (we fix `self.iter_per_epoch` to 1157 in this [line](https://github.com/dvlab-research/Context-Aware-Consistency/blob/master/train.py#L75)) in the fully-supervised setting. 

However, in the fully-supervised, `len(self.supervised_loader)` is actually a little larger than 1157. So, in fact, the number of training iterations in the fully-supervised setting should be a little larger. Even with fewer training steps, our model can achieve better performance, so I think it is considerably fair to make such comparison.

**Comment by ciuzaak** [comment]
Thanks for your answer.
I understand that while using all 10582 images in VOC dataset, `len(self.supervised_loader)` is larger than 1157.
Maybe what I described was not clear enough. I mainly want to know how to compare with SupOnly method. For example, with 300 labeled images, `len(self.supervised_loader)` is still less than your fixed iterations.

**Comment by X-Lai** [comment]
I guess you may misunderstand the config for SupOnly method.

You should note that for SupOnly method, the mode should still be 'semi' rather than 'supervised', so the epoch size is still `self.iter_per_epoch`. In this case, only the labeled data is used to compute the cross entropy loss, and the unlabeled data is simple discarded. Therefore, the comparison is absolutely fair.

**Comment by ciuzaak** [comment]
Ok I got it. Thank you very much!
